torch/testing/_asserts.py

```diff
@@ -577,31 +577,43 @@ def _to_tensor(array_or_scalar_like: Any) -> Tuple[Optional[_TestingErrorMeta],
     return None, tensor
 
 
-def _to_tensor_pair(actual: Any, expected: Any) -> Tuple[Optional[_TestingErrorMeta], Optional[_TensorPair]]:
+def _check_types(actual: Any, expected: Any, *, allow_subclasses: bool) -> Optional[_TestingErrorMeta]:
+    # We exclude numbers here, since numbers of different type, e.g. int vs. float, should be treated the same as
+    # tensors with different dtypes. Without user input, passing numbers of different types will still fail, but this
+    # can be disabled by setting `check_dtype=False`.
+    if isinstance(actual, numbers.Number) and isinstance(expected, numbers.Number):
+        return None
+
+    msg_fmtstr = f"Except for Python scalars, {{}}, but got {type(actual)} and {type(expected)} instead."
+    common_superclass = isinstance(actual, type(expected)) or isinstance(expected, type(actual))
+    if not common_superclass:
+        return _TestingErrorMeta(AssertionError, msg_fmtstr.format("inputs are required to share a common superclass"))
```

Comment:
This error message should reflect the value of `allow_subclasses`, too

If `allow_subclasses=False` the two objects must have the same type

If `allows_subclasses=True` then one object must be a subclass of the other's type

Reply:
I've only added the `allow_subclasses=False` in the error message, since "one object must be a subclass of the other's type" is true regardless of `allow_subclasses`.